File: lintcode/128-Hash-Function/solution.cpp
class Solution {
public:
    /**
     * @param key: A String you should hash
     * @param HASH_SIZE: An integer
     * @return an integer
     */
    int hashCode(string key,int HASH_SIZE) {
        if (key.size() == 0) return 0;

        long long result = 0;
        long long times = 1;
        for (int i = key.size() - 1; i >= 0; i--) {
            result += (int)key[i] * times % HASH_SIZE;
            times = times * 33 % HASH_SIZE;
        }
        
        return result % HASH_SIZE;
    }
};

File: lintcode/128-Hash-Function/solution.java
class Solution {
    /**
     * @param key: A String you should hash
     * @param HASH_SIZE: An integer
     * @return an integer
     */
    public int hashCode(char[] key,int HASH_SIZE) {
        if (key == null || key.length == 0) {
            return 0;
        }
        int number = 33;
        long curr = 1;
        long result = 0;
        for (int i = key.length - 1; i >= 0; i--) {
            result += curr * key[i] % HASH_SIZE;
            curr = (number * curr) % HASH_SIZE;
        }

        return (int) (result % HASH_SIZE);
    }
};